File: py/test/outbound_policy_test.py
# Copyright Modal Labs 2026
import pytest

from modal.exception import InvalidError
from modal.outbound_policy import _OutboundPolicy, _template_references_key
from modal.secret import _Secret


def test_template_references_key():
    assert _template_references_key("Bearer $API_KEY") is True
    assert _template_references_key("$_PRIVATE") is True
    assert _template_references_key("x$y") is True
    assert _template_references_key("plain") is False
    assert _template_references_key("ca$$h") is False
    assert _template_references_key("$$API_KEY") is False
    assert _template_references_key("$5") is False
    assert _template_references_key("$") is False
    assert _template_references_key("ends with $") is False
    assert _template_references_key("ca$$h $KEY") is True


def test_with_header_replacement_builds_immutable_policy():
    policy = _OutboundPolicy()
    updated = policy.with_header_replacement(domain="api.example.com", headers={"X-Static": "plain"})

    assert policy is not updated
    assert policy._replacements == ()
    assert len(updated._replacements) == 1

    replacement = updated._replacements[0]
    assert replacement.domain == "api.example.com"
    assert replacement.headers == {"X-Static": "plain"}
    assert replacement.secret is None


def test_to_proto_fans_out_replacements():
    policy = (
        _OutboundPolicy()
        .with_header_replacement(
            domain="api.example.com",
            headers={"X-First": "1", "X-Second": "2"},
        )
        .with_header_replacement(domain="*.example.com", headers={"X-Static": "plain"})
    )
    proto = policy._to_proto()
    assert len(proto.header_replacements) == 2

    first, second = proto.header_replacements
    assert first.domain == "api.example.com"
    assert first.secret_id == ""
    assert dict(first.headers) == {"X-First": "1", "X-Second": "2"}
    assert second.domain == "*.example.com"
    assert dict(second.headers) == {"X-Static": "plain"}


def test_validate_rejects_invalid_domain():
    with pytest.raises(InvalidError, match="Invalid domain"):
        _OutboundPolicy().with_header_replacement(domain="not a domain", headers={"a": "b"})._validate()

    with pytest.raises(InvalidError, match="Invalid domain"):
        _OutboundPolicy().with_header_replacement(domain="example.com\n", headers={"a": "b"})._validate()


def test_with_header_replacement_accepts_wildcards():
    policy = (
        _OutboundPolicy()
        .with_header_replacement(domain="*.example.com", headers={"a": "b"})
        .with_header_replacement(domain="*", headers={"a": "b"})
    )
    assert [s.domain for s in policy._replacements] == ["*.example.com", "*"]


def test_validate_rejects_empty_headers():
    with pytest.raises(InvalidError, match="at least one header"):
        _OutboundPolicy().with_header_replacement(domain="example.com", headers={})._validate()


def test_validate_rejects_invalid_header_name():
    with pytest.raises(InvalidError, match="Invalid header name"):
        _OutboundPolicy().with_header_replacement(domain="example.com", headers={"bad header": "x"})._validate()

    with pytest.raises(InvalidError, match="Invalid header name"):
        _OutboundPolicy().with_header_replacement(domain="example.com", headers={"X-Foo\n": "x"})._validate()


def test_validate_rejects_control_characters_in_value():
    for value in ["x\r\ninjected", "x\x00y", "x\x7fy"]:
        with pytest.raises(InvalidError, match="control characters"):
            _OutboundPolicy().with_header_replacement(domain="example.com", headers={"a": value})._validate()


def test_validate_allows_tab_in_value():
    _OutboundPolicy().with_header_replacement(domain="example.com", headers={"a": "x\ty"})._validate()


def test_validate_rejects_template_without_secret():
    with pytest.raises(InvalidError, match="no `secret` was passed"):
        _OutboundPolicy().with_header_replacement(
            domain="example.com", headers={"Authorization": "Bearer $API_KEY"}
        )._validate()


def test_validate_allows_escaped_dollar_without_secret():
    policy = _OutboundPolicy().with_header_replacement(
        domain="example.com", headers={"a": "ca$$h", "b": "$5", "c": "$"}
    )
    policy._validate()
    assert policy._replacements[0].headers == {"a": "ca$$h", "b": "$5", "c": "$"}


def test_validate_rejects_ephemeral_secret():
    with pytest.raises(InvalidError, match="only support named secrets"):
        _OutboundPolicy().with_header_replacement(
            domain="example.com",
            secret=_Secret.from_dict({"API_KEY": "k"}),
            headers={"Authorization": "Bearer $API_KEY"},
        )._validate()


def test_validate_rejects_too_many_headers():
    with pytest.raises(InvalidError, match="more than 25 header replacements"):
        _OutboundPolicy().with_header_replacement(
            domain="example.com", headers={f"X-Header-{i}": "x" for i in range(26)}
        )._validate()

    # The limit counts headers across replacements.
    policy = (
        _OutboundPolicy()
        .with_header_replacement(domain="example.com", headers={f"X-Header-{i}": "x" for i in range(25)})
        .with_header_replacement(domain="other.com", headers={"a": "b"})
    )
    with pytest.raises(InvalidError, match="more than 25 header replacements"):
        policy._validate()


def test_secrets_deduplicates():
    secret = _Secret.from_name("my-secret")
    policy = (
        _OutboundPolicy()
        .with_header_replacement(domain="a.example.com", secret=secret, headers={"a": "$K"})
        .with_header_replacement(domain="b.example.com", secret=secret, headers={"b": "$K"})
    )
    assert policy._secrets() == [secret]

File: js/test/outbound_policy.test.ts
import { expect, test } from "vitest";
import { InvalidError, OutboundPolicy, Secret } from "modal";
import { templateReferencesKey } from "../src/outbound_policy";

test("templateReferencesKey", () => {
  expect(templateReferencesKey("Bearer $API_KEY")).toBe(true);
  expect(templateReferencesKey("$_PRIVATE")).toBe(true);
  expect(templateReferencesKey("x$y")).toBe(true);
  expect(templateReferencesKey("plain")).toBe(false);
  expect(templateReferencesKey("ca$$h")).toBe(false);
  expect(templateReferencesKey("$$API_KEY")).toBe(false);
  expect(templateReferencesKey("$5")).toBe(false);
  expect(templateReferencesKey("$")).toBe(false);
  expect(templateReferencesKey("ends with $")).toBe(false);
  expect(templateReferencesKey("ca$$h $KEY")).toBe(true);
});

test("withHeaderReplacement builds immutable policy", () => {
  const policy = new OutboundPolicy();
  const updated = policy.withHeaderReplacement({
    domain: "api.example.com",
    headers: { "X-Static": "plain" },
  });

  expect(updated).not.toBe(policy);
  expect(policy._replacements).toEqual([]);
  expect(updated._replacements).toEqual([
    {
      domain: "api.example.com",
      headers: { "X-Static": "plain" },
      secret: undefined,
    },
  ]);
});

test("_toProto fans out replacements", () => {
  const secret = new Secret("st-1", "my-secret");
  const policy = new OutboundPolicy()
    .withHeaderReplacement({
      domain: "api.example.com",
      secret,
      headers: { Authorization: "Bearer $API_KEY", "X-Key-Raw": "$API_KEY" },
    })
    .withHeaderReplacement({
      domain: "*.example.com",
      headers: { "X-Static": "plain" },
    });

  const proto = policy._toProto();
  expect(proto.headerReplacements).toHaveLength(2);

  const [first, second] = proto.headerReplacements;
  expect(first.domain).toBe("api.example.com");
  expect(first.secretId).toBe("st-1");
  expect(first.headers).toEqual({
    Authorization: "Bearer $API_KEY",
    "X-Key-Raw": "$API_KEY",
  });
  expect(second.domain).toBe("*.example.com");
  expect(second.secretId).toBe("");
  expect(second.headers).toEqual({ "X-Static": "plain" });
});

test("_validate rejects invalid domain", () => {
  for (const domain of ["not a domain", "example.com\n"]) {
    const policy = new OutboundPolicy().withHeaderReplacement({
      domain,
      headers: { a: "b" },
    });
    expect(() => policy._validate()).toThrow(InvalidError);
  }
});

test("withHeaderReplacement accepts wildcards", () => {
  const policy = new OutboundPolicy()
    .withHeaderReplacement({ domain: "*.example.com", headers: { a: "b" } })
    .withHeaderReplacement({ domain: "*", headers: { a: "b" } });
  expect(policy._replacements.map((s) => s.domain)).toEqual([
    "*.example.com",
    "*",
  ]);
});

test("_validate rejects empty headers", () => {
  const policy = new OutboundPolicy().withHeaderReplacement({
    domain: "example.com",
    headers: {},
  });
  expect(() => policy._validate()).toThrow("at least one header");
});

test("_validate rejects invalid header name", () => {
  for (const name of ["bad header", "X-Foo\n"]) {
    const policy = new OutboundPolicy().withHeaderReplacement({
      domain: "example.com",
      headers: { [name]: "x" },
    });
    expect(() => policy._validate()).toThrow("Invalid header name");
  }
});

test("_validate rejects control characters in header value", () => {
  for (const value of ["x\r\ninjected", "x\x00y", "x\x7fy"]) {
    const policy = new OutboundPolicy().withHeaderReplacement({
      domain: "example.com",
      headers: { a: value },
    });
    expect(() => policy._validate()).toThrow("control characters");
  }
});

test("_validate allows tab in header value", () => {
  const policy = new OutboundPolicy().withHeaderReplacement({
    domain: "example.com",
    headers: { a: "x\ty" },
  });
  policy._validate();
  expect(policy._replacements[0].headers).toEqual({ a: "x\ty" });
});

test("_validate rejects template without secret", () => {
  const policy = new OutboundPolicy().withHeaderReplacement({
    domain: "example.com",
    headers: { Authorization: "Bearer $API_KEY" },
  });
  expect(() => policy._validate()).toThrow("no `secret` was passed");
});

test("_validate allows escaped dollar without secret", () => {
  const policy = new OutboundPolicy().withHeaderReplacement({
    domain: "example.com",
    headers: { a: "ca$$h", b: "$5", c: "$" },
  });
  policy._validate();
  expect(policy._replacements[0].headers).toEqual({
    a: "ca$$h",
    b: "$5",
    c: "$",
  });
});

test("_validate rejects too many headers", () => {
  const headers: Record<string, string> = {};
  for (let i = 0; i < 26; i++) headers[`X-Header-${i}`] = "x";
  const policy = new OutboundPolicy().withHeaderReplacement({
    domain: "example.com",
    headers,
  });
  expect(() => policy._validate()).toThrow("more than 25 header replacements");

  // The limit counts headers across replacements.
  const full: Record<string, string> = {};
  for (let i = 0; i < 25; i++) full[`X-Header-${i}`] = "x";
  const overLimit = new OutboundPolicy()
    .withHeaderReplacement({ domain: "example.com", headers: full })
    .withHeaderReplacement({ domain: "other.com", headers: { a: "b" } });
  expect(() => overLimit._validate()).toThrow(
    "more than 25 header replacements",
  );
});

test("_secrets deduplicates", () => {
  const secret = new Secret("st-1", "my-secret");
  const policy = new OutboundPolicy()
    .withHeaderReplacement({
      domain: "a.example.com",
      secret,
      headers: { a: "$K" },
    })
    .withHeaderReplacement({
      domain: "b.example.com",
      secret,
      headers: { b: "$K" },
    });
  expect(policy._secrets()).toEqual([secret]);
});

test("_validate rejects ephemeral secrets", async () => {
  const { createMockModalClients } = await import("../test-support/grpc_mock");
  const { mockClient: mc } = createMockModalClients();
  const ephemeral = await mc.secrets.fromObject({ API_KEY: "k" });
  const policy = new OutboundPolicy().withHeaderReplacement({
    domain: "example.com",
    secret: ephemeral,
    headers: { Authorization: "Bearer $API_KEY" },
  });
  expect(() => policy._validate()).toThrow("only support named secrets");
});
